code/iaas/logic/src/main/java/io/cattle/platform/process/externalevent/ExternalEventCreate.java

```diff
@@ -0,0 +1,209 @@
+
+        if (StringUtils.isEmpty(event.getExternalId())) {
+            log.warn("External event doesn't have an external id: {}", event);
+            return null;
+        }
+
+        if (ExternalEventConstants.KIND_VOLUME_EVENT.equals(event.getKind())) {
+            handleVolumeEvent(event, state, process);
+        } else if (ExternalEventConstants.KIND_STORAGE_POOL_EVENT.equals(event.getKind())) {
+            handleStoragePoolEvent(event, state, process);
+        } else {
+            new IllegalStateException();
+        }
+
+        return null;
+    }
+
+    protected void handleVolumeEvent(final ExternalEvent event, ProcessState state, ProcessInstance process) {
+        lockManager.lock(new ExternalEventLock(VOLUME_POOL_LOCK_NAME, event.getAccountId(), event.getExternalId()), new LockCallbackNoReturn() {
+            @Override
+            public void doWithLockNoResult() {
+                String spExtId = DataAccessor.fieldString(event, FIELD_SP_EXT_ID);
+                StoragePool storagePool = storagePoolDao.findStoragePoolByExternalId(event.getAccountId(), spExtId);
+                if (storagePool == null) {
+                    log.warn("Unknown storage pool. Returning. External id: {}", spExtId);
+                    return;
+                }
+                Volume volume = volumeDao.findVolumeByExternalId(storagePool.getId(), event.getExternalId());
+                switch (event.getEventType()) {
+                case ExternalEventConstants.TYPE_VOLUME_CREATE:
+                    if (volume == null) {
+                        Map<String, Object> volumeData = CollectionUtils.toMap(DataUtils.getFields(event).get("volume"));
+                        if (volumeData == null || volumeData.isEmpty()) {
+                            log.warn("Null or empty volume for externalVolumeEvent: {}. StoragePool: {}", event, volumeData);
+                            return;
+                        }
+
+                        volumeData.put(ObjectMetaDataManager.ACCOUNT_FIELD, event.getAccountId());
+                        volumeData.put(FIELD_ATTACHED_STATE, CommonStatesConstants.INACTIVE);
+                        volumeData.put(FIELD_ALLOC_STATE, CommonStatesConstants.ACTIVE);
+                        volumeData.put(FIELD_ZONE_ID, 1L);
+                        volumeData.put(FIELD_DEV_NUM, -1);
+
+                        volumeDao.createVolumeInStoragePool(volumeData, storagePool);
+                    }
+                    break;
+                case ExternalEventConstants.TYPE_VOLUME_DELETE:
+                    if (volume != null) {
+                        try {
+                            objectProcessManager.scheduleStandardProcess(StandardProcess.DEACTIVATE, volume,
+                                    ProcessUtils.chainInData(new HashMap<String, Object>(), PROC_VOL_DEACTIVATE, PROC_VOL_REMOVE));
+                        } catch (ProcessCancelException e) {
```

Comment:
I catch and ignore here because these external events run in the background without the user having a way address ones that end up continually rescheduling.

So, the idea is that this makes a "best effort" attempt at scheduling the process for volume or storage pool. We do the same thing with container events.

In a future PR, we'll add sync/ping logic similar to what we have for containers that runs periodically to make sure things sync up.

I do believe it is worth adding an info level log statement though.

Seem reasonable?